# Download human data

This short markdown file documents the downloading of the human data for the project.


```bash
#chromosomal data
wget ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/supporting/GRCh38_positions/ALL.chr1_GRCh38.genotypes.20170504.vcf.gz
wget ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/supporting/GRCh38_positions/ALL.chr1_GRCh38.genotypes.20170504.vcf.gz.tbi
wget ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/supporting/GRCh38_positions/ALL.chr2_GRCh38.genotypes.20170504.vcf.gz
wget ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/supporting/GRCh38_positions/ALL.chr2_GRCh38.genotypes.20170504.vcf.gz.tbi
wget ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/supporting/GRCh38_positions/ALL.chr3_GRCh38.genotypes.20170504.vcf.gz
wget ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/supporting/GRCh38_positions/ALL.chr3_GRCh38.genotypes.20170504.vcf.gz.tbi
wget ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/supporting/GRCh38_positions/ALL.chr4_GRCh38.genotypes.20170504.vcf.gz
wget ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/supporting/GRCh38_positions/ALL.chr4_GRCh38.genotypes.20170504.vcf.gz.tbi
wget ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/supporting/GRCh38_positions/ALL.chr5_GRCh38.genotypes.20170504.vcf.gz
wget ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/supporting/GRCh38_positions/ALL.chr5_GRCh38.genotypes.20170504.vcf.gz.tbi
wget ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/supporting/GRCh38_positions/ALL.chr6_GRCh38.genotypes.20170504.vcf.gz
wget ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/supporting/GRCh38_positions/ALL.chr6_GRCh38.genotypes.20170504.vcf.gz.tbi
wget ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/supporting/GRCh38_positions/ALL.chr7_GRCh38.genotypes.20170504.vcf.gz
wget ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/supporting/GRCh38_positions/ALL.chr7_GRCh38.genotypes.20170504.vcf.gz.tbi
wget ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/supporting/GRCh38_positions/ALL.chr8_GRCh38.genotypes.20170504.vcf.gz
wget ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/supporting/GRCh38_positions/ALL.chr8_GRCh38.genotypes.20170504.vcf.gz.tbi
wget ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/supporting/GRCh38_positions/ALL.chr9_GRCh38.genotypes.20170504.vcf.gz
wget ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/supporting/GRCh38_positions/ALL.chr9_GRCh38.genotypes.20170504.vcf.gz.tbi
wget ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/supporting/GRCh38_positions/ALL.chr10_GRCh38.genotypes.20170504.vcf.gz
wget ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/supporting/GRCh38_positions/ALL.chr10_GRCh38.genotypes.20170504.vcf.gz.tbi
wget ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/supporting/GRCh38_positions/ALL.chr11_GRCh38.genotypes.20170504.vcf.gz
wget ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/supporting/GRCh38_positions/ALL.chr11_GRCh38.genotypes.20170504.vcf.gz.tbi
wget ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/supporting/GRCh38_positions/ALL.chr12_GRCh38.genotypes.20170504.vcf.gz
wget ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/supporting/GRCh38_positions/ALL.chr12_GRCh38.genotypes.20170504.vcf.gz.tbi
wget ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/supporting/GRCh38_positions/ALL.chr13_GRCh38.genotypes.20170504.vcf.gz
wget ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/supporting/GRCh38_positions/ALL.chr13_GRCh38.genotypes.20170504.vcf.gz.tbi
wget ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/supporting/GRCh38_positions/ALL.chr14_GRCh38.genotypes.20170504.vcf.gz
wget ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/supporting/GRCh38_positions/ALL.chr14_GRCh38.genotypes.20170504.vcf.gz.tbi
wget ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/supporting/GRCh38_positions/ALL.chr15_GRCh38.genotypes.20170504.vcf.gz
wget ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/supporting/GRCh38_positions/ALL.chr15_GRCh38.genotypes.20170504.vcf.gz.tbi
wget ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/supporting/GRCh38_positions/ALL.chr16_GRCh38.genotypes.20170504.vcf.gz
wget ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/supporting/GRCh38_positions/ALL.chr16_GRCh38.genotypes.20170504.vcf.gz.tbi
wget ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/supporting/GRCh38_positions/ALL.chr17_GRCh38.genotypes.20170504.vcf.gz
wget ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/supporting/GRCh38_positions/ALL.chr17_GRCh38.genotypes.20170504.vcf.gz.tbi
wget ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/supporting/GRCh38_positions/ALL.chr18_GRCh38.genotypes.20170504.vcf.gz
wget ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/supporting/GRCh38_positions/ALL.chr18_GRCh38.genotypes.20170504.vcf.gz.tbi
wget ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/supporting/GRCh38_positions/ALL.chr19_GRCh38.genotypes.20170504.vcf.gz
wget ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/supporting/GRCh38_positions/ALL.chr19_GRCh38.genotypes.20170504.vcf.gz.tbi
wget ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/supporting/GRCh38_positions/ALL.chr20_GRCh38.genotypes.20170504.vcf.gz
wget ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/supporting/GRCh38_positions/ALL.chr20_GRCh38.genotypes.20170504.vcf.gz.tbi
wget ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/supporting/GRCh38_positions/ALL.chr21_GRCh38.genotypes.20170504.vcf.gz
wget ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/supporting/GRCh38_positions/ALL.chr21_GRCh38.genotypes.20170504.vcf.gz.tbi
wget ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/supporting/GRCh38_positions/ALL.chr22_GRCh38.genotypes.20170504.vcf.gz
wget ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/supporting/GRCh38_positions/ALL.chr22_GRCh38.genotypes.20170504.vcf.gz.tbi
```

Now onto concatenation, cleanly for oder

```bash
zcat ALL.chr1_GRCh38.genotypes.20170504.vcf.gz | head -n 10000 | grep "#" > headers.txt 

zcat ALL.chr1_GRCh38.genotypes.20170504.vcf.gz ALL.chr2_GRCh38.genotypes.20170504.vcf.gz ALL.chr3_GRCh38.genotypes.20170504.vcf.gz ALL.chr4_GRCh38.genotypes.20170504.vcf.gz ALL.chr5_GRCh38.genotypes.20170504.vcf.gz ALL.chr6_GRCh38.genotypes.20170504.vcf.gz ALL.chr7_GRCh38.genotypes.20170504.vcf.gz ALL.chr8_GRCh38.genotypes.20170504.vcf.gz ALL.chr9_GRCh38.genotypes.20170504.vcf.gz ALL.chr10_GRCh38.genotypes.20170504.vcf.gz ALL.chr11_GRCh38.genotypes.20170504.vcf.gz ALL.chr12_GRCh38.genotypes.20170504.vcf.gz ALL.chr13_GRCh38.genotypes.20170504.vcf.gz ALL.chr14_GRCh38.genotypes.20170504.vcf.gz ALL.chr15_GRCh38.genotypes.20170504.vcf.gz ALL.chr16_GRCh38.genotypes.20170504.vcf.gz ALL.chr17_GRCh38.genotypes.20170504.vcf.gz ALL.chr18_GRCh38.genotypes.20170504.vcf.gz ALL.chr19_GRCh38.genotypes.20170504.vcf.gz ALL.chr20_GRCh38.genotypes.20170504.vcf.gz ALL.chr21_GRCh38.genotypes.20170504.vcf.gz ALL.chr22_GRCh38.genotypes.20170504.vcf.gz | grep -v "#" | grep -v INDEL | cat headers.txt - | gzip > allchrom.vcf.gz
```

I Just realised that was not bgzippied just gzipped:


```bash
#module load SAMtools
 gunzip -c  allchrom.vcf.gz  | bgzip -c > allchrom2.vcf.gz
 tabix allchrom2.vcf.gz
```

## Annotation
 make an exon one and a cds annotation file.

```bash
wget ftp://ftp.ensembl.org/pub/release-99/gtf/homo_sapiens/Homo_sapiens.GRCh38.99.chr.gtf.gz
gunzip Homo_sapiens.GRCh38.99.chr.gtf.gz
```

Make the CDS only annotation file

```python
#python

outputCDS=open("Homo_sapiens.GRCh38.99.CDS.chr.gtf","w")

with open("Homo_sapiens.GRCh38.99.chr.gtf") as f:
	for line in f:
		if not line.startswith("#"):
			if line.split()[2] == "CDS": outputCDS.write(line)
outputCDS.close()
```
Annotation is ready!


## Extract the vcf files for CDS only

merge redundant features:

```
sort -k1,1 -k2,2n Homo_sapiens.GRCh38.99.CDS.chr.gtf > Homo_sapiens.GRCh38.99.CDS.chr.sorted.gtf #useless need to resort later
 cut -f 1,4,5 Homo_sapiens.GRCh38.99.CDS.chr.sorted.gtf | awk -F " " '{print $1 "\t" $2-1 "\t" $3}' > Homo_sapiens.GRCh38.99.CDS.chr.sorted.bed

 sort -k1,1 -k2,2n Homo_sapiens.GRCh38.99.CDS.chr.sorted.bed > Homo_sapiens.GRCh38.99.CDS.chr.sorted.sorted.bed
bedtools merge -i Homo_sapiens.GRCh38.99.CDS.chr.sorted.sorted.bed > Homo_sapiens.GRCh38.99.CDS.chr.cleaned.bed
```

the clean file is Homo_sapiens.GRCh38.99.CDS.chr.cleaned.bed


```python
import os
i=0

with open("Homo_sapiens.GRCh38.99.CDS.chr.cleaned.bed") as f:
	for line in f:
		scaf,start,end = line.split()[:3]
		os.system("tabix allchrom2.vcf.gz "+scaf+":"+start+"-"+end+" >> allcds_toclean.vcf")# the plus one is bed 0 based to vcf 1 based

```
```bash
module load BCFtools
zcat allchrom2.vcf.gz  | head -n 100000 | grep "^#" | cat - allexons_toclean.vcf | bcftools sort - | uniq > 1000humans_onlycds_clean.vcf
```

I also grabbed metadata from [https://www.internationalgenome.org/data-portal/sample](https://www.internationalgenome.org/data-portal/sample)

and put into : [igsr_samples.tsv](igsr_samples.tsv)


